tox_core/src/dht/forced_ktree.rs

```diff
@@ -0,0 +1,438 @@
+    ktree: Ktree,
+    kbucket: Kbucket<DhtNode>,
+}
+
+impl ForcedKtree {
+    /// Create a new `ForceKtree`.
+    pub fn new(pk: &PublicKey) -> Self {
+        ForcedKtree {
+            ktree: Ktree::new(pk),
+            kbucket: Kbucket::new(KBUCKET_DEFAULT_SIZE),
+        }
+    }
+
+    /// Get reference to a `DhtNode` by it's `PublicKey`.
+    pub fn get_node(&self, pk: &PublicKey) -> Option<&DhtNode> {
+        self.ktree.get_node(pk).or_else(||
+            self.kbucket.get_node(&self.ktree.pk, pk)
+        )
+    }
+
+    /// Get mutable reference to a `DhtNode` by it's `PublicKey`.
+    pub fn get_node_mut(&mut self, pk: &PublicKey) -> Option<&mut DhtNode> {
+        let base_pk = self.ktree.pk;
+        let bucket = &mut self.kbucket;
+        self.ktree.get_node_mut(pk).or_else(move ||
+            bucket.get_node_mut(&base_pk, pk)
+        )
+    }
+
+    /** Add [`PackedNode`](./struct.PackedNode.html) to `Kbucket`.
+
+    Node can be added only if:
+
+    * its [`kbucket index`](./fn.kbucket_index.html) is lower than the
+      number of buckets.
+    * [`Bucket`](./struct.Bucket.html) to which it is added has free space
+      or added node is closer to the PK than other node in the bucket.
+
+    Returns `true` if node was added successfully, `false` otherwise.
+    */
+    pub fn try_add(&mut self, node: PackedNode) -> bool {
+        if self.ktree.try_add(node) {
+            if let Some(dht_node) = self.kbucket.remove(&self.ktree.pk, &node.pk) {
+                let added_node = self.ktree.get_node_mut(&node.pk).expect("Node should be added");
+                if node.saddr.is_ipv4() {
+                    added_node.assoc6 = dht_node.assoc6;
+                } else {
+                    added_node.assoc4 = dht_node.assoc4;
+                }
+            }
+            true
+        } else if !self.ktree.contains(&node.pk) {
+            self.kbucket.try_add(&self.ktree.pk, node, /* evict */ true)
+        } else {
+            false
+        }
+    }
+
+    /// Remove [`DhtNode`](./struct.DhtNode.html) with given PK from the
+    /// `Kbucket`.
+    pub fn remove(&mut self, node_pk: &PublicKey) -> Option<DhtNode> {
+        self.ktree.remove(node_pk).or_else(||
+            self.kbucket.remove(&self.ktree.pk, node_pk)
+        )
+    }
+
+    /** Get (up to) `count` closest nodes to given PK.
+
+    Functionality for [`SendNodes`](./struct.SendNodes.html).
+
+    Returns less than `count` nodes only if `Ktree` contains less than `count`
+    nodes.
+
+    It should not contain LAN ip node if the request is from global ip.
+    */
+    pub fn get_closest(&self, pk: &PublicKey, count: u8, only_global: bool) -> Kbucket<PackedNode> {
+        let mut kbucket = self.ktree.get_closest(pk, count, only_global);
+        for node in self.kbucket.iter().filter(|node| !node.is_bad()) {
+            if let Some(pn) = node.to_packed_node() {
+                if !only_global || IsGlobal::is_global(&pn.saddr.ip()) {
```

Comment:
Is it a correct check?

Reply:
Yes, it's either global and we add it, or local and we include not only global so we still add it.